File: src/ui/lcd/mute-edit.ts
/**
 * LCD Mute Edit screen — shows all 4 tracks' mute patterns at once.
 * Selected track is highlighted; step buttons toggle mutes on the selected track.
 * Fixed 16-step view (no paging), all tracks visible for arrangement overview.
 */

import type { SequencerState } from '../../engine/types'
import type { UIState } from '../hw-types'
import { COLORS } from '../colors'
import { fillRect, strokeRect, drawText, LCD_W, LCD_CONTENT_Y, LCD_CONTENT_H } from '../renderer'

const PAD = 8
const LABEL_W = 28
const HEADER_H = 24
const TRACK_GAP = 4
const COLS = 16
const COL_GAP = 2
const GRID_LEFT = PAD + LABEL_W
const STEP_W = (LCD_W - GRID_LEFT - PAD - (COLS - 1) * COL_GAP) / COLS
const AVAIL_H = LCD_CONTENT_H - HEADER_H - 4
const ROW_H = (AVAIL_H - 3 * TRACK_GAP) / 4

const MUTE_COLOR = '#aa3344'

export function renderMuteEdit(ctx: CanvasRenderingContext2D, engine: SequencerState, ui: UIState): void {
  // Title
  drawText(ctx, 'MUTE', PAD, LCD_CONTENT_Y + 18, COLORS.textBright, 18)

  // Render all 4 tracks
  for (let t = 0; t < 4; t++) {
    const mute = engine.mutePatterns[t]
    const trackColor = COLORS.track[t]
    const dimColor = COLORS.trackDim[t]
    const isSelected = t === ui.selectedTrack
    const rowY = LCD_CONTENT_Y + HEADER_H + t * (ROW_H + TRACK_GAP)

    // Selected track background highlight
    if (isSelected) {
      fillRect(ctx, { x: PAD, y: rowY - 1, w: LCD_W - PAD * 2, h: ROW_H + 2 }, `${trackColor}20`)
    }

    // Track label
    drawText(ctx, `T${t + 1}`, PAD + 2, rowY + ROW_H / 2 + 1, isSelected ? trackColor : COLORS.textDim, 16)

    // Mute step cells
    for (let i = 0; i < COLS; i++) {
      const x = GRID_LEFT + i * (STEP_W + COL_GAP)

      if (i >= mute.length) {
        fillRect(ctx, { x, y: rowY, w: STEP_W, h: ROW_H }, '#111118')
      } else if (mute.steps[i]) {
        // Muted
        fillRect(ctx, { x, y: rowY, w: STEP_W, h: ROW_H }, isSelected ? MUTE_COLOR : `${MUTE_COLOR}88`)
      } else {
        // Not muted
        fillRect(ctx, { x, y: rowY, w: STEP_W, h: ROW_H }, isSelected ? dimColor : `${dimColor}60`)
      }

      // Playhead
      if (i === mute.currentStep) {
        strokeRect(ctx, { x: x - 1, y: rowY - 1, w: STEP_W + 2, h: ROW_H + 2 }, '#ffffff', isSelected ? 2 : 1)
      }
    }
  }
}

File: crates/renderer/src/screens/mute_edit.rs
use embedded_graphics::{pixelcolor::Rgb565, prelude::*};
use requencer_engine::types::SequencerState;

use crate::{colors, draw, layout, types::UiState};

/// Render the mute edit screen: 4 rows x 16 columns showing all mute patterns.
pub fn render<D: DrawTarget<Color = Rgb565>>(
    display: &mut D,
    state: &SequencerState,
    ui: &UiState,
) {
    // Header
    draw::fill_rect(
        display,
        0,
        layout::CONTENT_Y as i32,
        layout::LCD_W,
        layout::EDIT_HEADER_H,
        colors::STATUS_BAR,
    );
    draw::text(
        display,
        layout::PAD as i32,
        layout::CONTENT_Y as i32 + 9,
        "MUTE",
        colors::TRACK[ui.selected_track as usize],
    );

    let grid_y = layout::CONTENT_Y as i32 + layout::EDIT_HEADER_H as i32;
    let avail_h = layout::CONTENT_H - layout::EDIT_HEADER_H - layout::EDIT_FOOTER_H;
    let row_h = avail_h / 4;
    let step_w = (layout::LCD_W - 28) / 16;

    for track_idx in 0..4usize {
        let mute = &state.mute_patterns[track_idx];
        let color = colors::TRACK[track_idx];
        let dim = colors::TRACK_DIM[track_idx];
        let is_selected_track = ui.selected_track as usize == track_idx;
        let y = grid_y + track_idx as i32 * row_h as i32;

        // Selected track row highlight
        if is_selected_track {
            draw::fill_rect(display, 0, y, layout::LCD_W, row_h, dim);
        }

        // Track label
        let label = match track_idx {
            0 => "T1",
            1 => "T2",
            2 => "T3",
            _ => "T4",
        };
        draw::text(
            display,
            2,
            y + (row_h as i32 / 2) - 5,
            label,
            if is_selected_track { color } else { colors::TEXT_DIM },
        );

        // Mute steps
        for i in 0..mute.length.min(16) as usize {
            let x = 24 + i as i32 * step_w as i32;
            let is_playhead = mute.current_step as usize == i;
            let muted = i < mute.steps.len() && mute.steps[i];
            let cell_h = row_h - 4;
            let cy = y + 2;

            let c = if muted {
                colors::STEP_MUTED
            } else if is_selected_track {
                color
            } else {
                dim
            };
            draw::fill_rect(display, x + 1, cy, step_w - 2, cell_h, c);

            // Playhead outline stroke
            if is_playhead {
                draw::stroke_rect(display, x, cy - 1, step_w, cell_h + 2, colors::PLAYHEAD);
            }
        }

        // Separator
        if track_idx < 3 {
            let sep_y = y + row_h as i32 - 1;
            draw::fill_rect(display, 0, sep_y, layout::LCD_W, 1, colors::STEP_OFF);
        }
    }

    // Footer
    let footer_y = layout::LCD_H as i32 - layout::EDIT_FOOTER_H as i32;
    draw::fill_rect(
        display,
        0,
        footer_y,
        layout::LCD_W,
        layout::EDIT_FOOTER_H,
        colors::STATUS_BAR,
    );
}
